subtests/docker_cli/login/login.py

```diff
@@ -0,0 +1,333 @@
+        at the beginning, then fail with HTTP 401 (Unauthorized). Once
+        we get that, trust that it's ready for login.
+        """
+        if timeout is None:
+            timeout = 10
+
+        url = "https://{server}/v2/".format(**self.sub_stuff)
+        end_time = time.time() + timeout
+        while time.time() <= end_time:
+            try:
+                urllib2.urlopen(url)
+            except urllib2.HTTPError as e:
+                if e.reason == 'Unauthorized':
+                    return
+            except urllib2.URLError as e:
+                # FIXME: assume that it's SSLEOFError; how can we make sure?
+                pass
+            time.sleep(0.5)
+        raise DockerTestFail("timed out waiting for registry")
+
+    def docker_login(self, password=None):
+        """
+        Run docker login against our local registry, using the supplied
+        password (default: correct password for our setup).
+        """
+        if password is None:
+            password = self.sub_stuff['passwd']
+        dc = DockerCmd(self, 'login', ['--username', self.sub_stuff['user'],
+                                       '--password', password,
+                                       '--email', 'nobody@redhat.com',
+                                       self.sub_stuff['server']])
+        return dc.execute()
+
+    def _check_credentials(self):
+        """
+        Examine saved credentials in docker auth file.
+        If this is a logout test, make sure the server key is absent.
+        Otherwise, make sure there's an auth entry for this server,
+        and its value contains the expected username + password.
+        """
+        auth_file = self.config['docker_auth_file']
+        fp = open(auth_file, 'rb')
+        creds = json.load(fp)
+        server = self.sub_stuff['server']
+
+        # Use test name to determine whether cred should be present or not
+        if 'logout' in self.__class__.__name__:
+            self.failif(server in creds['auths'],
+                        ('"%s" should not be in %s after docker logout'
+                         % (server, auth_file)))
+        else:
+            # Stored credentials
+            cred_hash = creds['auths'][server]['auth']
+            # ...simply a base64-encoded('username:password'), see:
+            #  https://coreos.com/os/docs/latest/registry-authentication.html
+            expected = b64encode('{user}:{passwd}'.format(**self.sub_stuff))
+            self.failif_ne(cred_hash, expected,
+                           'saved credentials in %s' % auth_file)
+
+    def docker_push(self, img_name, tag_name):
+        """
+        Run docker tag, then docker push against our local registry.
+        Tag should always succeed but make no assumptions about push.
+        """
+        push_image = self.config['push_image']
+        dc = DockerCmd(self, 'pull', [push_image])
+        mustpass(dc.execute())
+        self.sub_stuff['my_images'] = [push_image]
```

Comment:
Did you mean ``self.sub_stuff['my_images'] += [push_image]`` or was ``my_images`` not initialized to a list yet?